v-next/core/src/internal/tasks/resolved-task.ts

```diff
@@ -69,24 +77,190 @@ export class ResolvedTask implements Task {
+    // been processed. If there are any extra parameters, an error is thrown
+    this.#validateExtraArguments(providedArgumentNames);
+
+    const next = async (
+      nextTaskArguments: TaskArguments,
+      currentIndex = this.actions.length - 1,
+    ): Promise<any> => {
+      // The first action may be empty if the task was originally an empty task
+      const currentAction = this.actions[currentIndex].action ?? (() => {});
+
+      const actionFn =
+        typeof currentAction === "function"
+          ? currentAction
+          : await this.#resolveFileAction(currentAction, this.id);
+
+      if (currentIndex === 0) {
+        /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions --
+        We know that the first action in the array is a NewTaskActionFunction */
+        return (actionFn as NewTaskActionFunction)(
+          nextTaskArguments,
+          this.#hre,
         );
-      } else {
-        console.log(`  Running action: ${action.action?.toString()}`);
       }
+
+      return actionFn(
+        nextTaskArguments,
+        this.#hre,
+        async (newTaskArguments: TaskArguments) => {
+          return next(newTaskArguments, currentIndex - 1);
+        },
+      );
+    };
+
+    return next(taskArguments);
+  }
+
+  /**
+   * Validates that a required parameter has a value. A parameter is required if
+   * it doesn't have a default value.
+   *
+   * @throws HardhatError if the parameter is required and doesn't have a value.
+   */
+  #validateRequiredParameter(
+    parameter: TaskParameter,
+    value: ParameterValue | ParameterValue[],
+  ) {
+    if (parameter.defaultValue === undefined && value === undefined) {
+      throw new HardhatError(
+        HardhatError.ERRORS.ARGUMENTS.MISSING_VALUE_FOR_PARAMETER,
+        {
+          paramName: parameter.name,
+        },
+      );
+    }
+  }
+
+  /**
+   * Validates that a parameter has the correct type. If the parameter is optional
+   * and doesn't have a value, the type is not validated as it will be resolved
+   * to the default value.
+   *
+   * @throws HardhatError if the parameter has an invalid type.
+   */
+  #validateParameterType(
+    parameter: TaskParameter,
+    value: ParameterValue | ParameterValue[],
+  ) {
+    // skip type validation for optional parameters with undefined value
+    if (value === undefined && parameter.defaultValue !== undefined) {
+      return;
+    }
+
+    // check if the parameter is variadic
+    const isPositionalParameter = (
+      param: TaskParameter,
+    ): param is PositionalTaskParameter => "isVariadic" in param;
+    const isVariadic = isPositionalParameter(parameter) && parameter.isVariadic;
+
+    // check if the value is valid for the parameter type
+    if (!isParameterValueValid(parameter.parameterType, value, isVariadic)) {
+      throw new HardhatError(
+        HardhatError.ERRORS.ARGUMENTS.INVALID_VALUE_FOR_TYPE,
+        {
+          value,
+          name: parameter.name,
```

Comment:
Again, i think adding the task name would be valuable

Reply:
Same here